File: digimon_gym/engine/data/scripts/st19/st19_11.py
from __future__ import annotations
from typing import TYPE_CHECKING, List, Dict, Any
from ....core.card_script import CardScript
from ....interfaces.card_effect import ICardEffect
from ....data.enums import EffectTiming

if TYPE_CHECKING:
    from ....core.card_source import CardSource


class ST19_11(CardScript):
    """ST19-11 Chaperomon | Lv.5 (Yellow, Puppet/LIBERATOR)

    [On Play] [When Digivolving] 1 of your opponent's Digimon gets -3000 DP
    for the turn. If there are 3 or more Digimon, increase the DP reduction
    of this effect by -3000.

    Inherited Effect:
    [All Turns] [Once Per Turn] When this Digimon would leave the battle area
    other than by your effects, by deleting 1 of your Tokens or other [Puppet]
    trait Digimon, prevent it from leaving.
    """

    def get_card_effects(self, card: 'CardSource') -> List['ICardEffect']:
        effects = []

        def _get_dp_reduction() -> int:
            """Calculate DP reduction: -3000, or -6000 if 3+ total Digimon."""
            owner = card.owner if card else None
            if not owner:
                return -3000
            own_digimon = sum(1 for p in owner.battle_area if p.is_digimon)
            enemy = owner.enemy
            opp_digimon = sum(1 for p in enemy.battle_area if p.is_digimon) if enemy else 0
            total = own_digimon + opp_digimon
            if total >= 3:
                return -6000
            return -3000

        # --- Effect 0: [On Play] Opponent's Digimon gets -3000/-6000 DP ---
        effect0 = ICardEffect()
        effect0.set_timing(EffectTiming.OnEnterFieldAnyone)
        effect0.set_effect_name("ST19-11 On Play: DP reduction")
        effect0.set_effect_description(
            "[On Play] 1 of your opponent's Digimon gets -3000 DP for the "
            "turn. If there are 3 or more Digimon, increase the DP reduction "
            "of this effect by -3000."
        )
        effect0.is_on_play = True

        def condition0(context: Dict[str, Any]) -> bool:
            if card and card.permanent_of_this_card() is None:
                return False
            owner = card.owner if card else None
            if not owner:
                return False
            enemy = owner.enemy
            if not enemy:
                return False
            return any(p.is_digimon for p in enemy.battle_area)

        effect0.set_can_use_condition(condition0)

        def process0(ctx: Dict[str, Any]):
            """Select 1 opponent Digimon and apply DP reduction."""
            player = ctx.get('player')
            game = ctx.get('game')
            if not (player and game):
                return
            reduction = _get_dp_reduction()

            def on_select(target_perm):
                target_perm.change_dp(reduction)

            game.effect_select_opponent_permanent(
                player, on_select,
                filter_fn=lambda p: p.is_digimon,
                is_optional=False,
                prompt=f"Select 1 opponent Digimon to get {reduction} DP.",
            )

        effect0.set_on_process_callback(process0)
        effects.append(effect0)

        # --- Effect 1: [When Digivolving] Same DP reduction ---
        effect1 = ICardEffect()
        effect1.set_timing(EffectTiming.OnEnterFieldAnyone)
        effect1.set_effect_name("ST19-11 When Digivolving: DP reduction")
        effect1.set_effect_description(
            "[When Digivolving] 1 of your opponent's Digimon gets -3000 DP "
            "for the turn. If there are 3 or more Digimon, increase the DP "
            "reduction of this effect by -3000."
        )
        effect1.is_when_digivolving = True

        def condition1(context: Dict[str, Any]) -> bool:
            if card and card.permanent_of_this_card() is None:
                return False
            owner = card.owner if card else None
            if not owner:
                return False
            enemy = owner.enemy
            if not enemy:
                return False
            return any(p.is_digimon for p in enemy.battle_area)

        effect1.set_can_use_condition(condition1)

        def process1(ctx: Dict[str, Any]):
            """Select 1 opponent Digimon and apply DP reduction."""
            player = ctx.get('player')
            game = ctx.get('game')
            if not (player and game):
                return
            reduction = _get_dp_reduction()

            def on_select(target_perm):
                target_perm.change_dp(reduction)

            game.effect_select_opponent_permanent(
                player, on_select,
                filter_fn=lambda p: p.is_digimon,
                is_optional=False,
                prompt=f"Select 1 opponent Digimon to get {reduction} DP.",
            )

        effect1.set_on_process_callback(process1)
        effects.append(effect1)

        # --- Effect 2 (Inherited): [All Turns] [Once Per Turn] Prevent leaving ---
        # When this Digimon would leave the battle area other than by your
        # effects, by deleting 1 of your Tokens or other [Puppet] trait
        # Digimon, prevent it from leaving.
        effect2 = ICardEffect()
        effect2.set_timing(EffectTiming.WhenRemoveField)
        effect2.set_effect_name("ST19-11 Prevent leaving by deleting Token/Puppet")
        effect2.set_effect_description(
            "[All Turns] [Once Per Turn] When this Digimon would leave the "
            "battle area other than by your effects, by deleting 1 of your "
            "Tokens or other [Puppet] trait Digimon, prevent it from leaving."
        )
        effect2.is_inherited_effect = True
        effect2.is_optional = True
        effect2.set_max_count_per_turn(1)
        effect2.set_hash_string("Substitute_EX7_027")

        def condition2(context: Dict[str, Any]) -> bool:
            if card and card.permanent_of_this_card() is None:
                return False
            owner = card.owner if card else None
            if not owner:
                return False
            my_perm = card.permanent_of_this_card()
            # The leaving permanent must be this Digimon
            leaving_perm = context.get('permanent')
            if leaving_perm is not my_perm:
                return False
            # Must not be caused by own effects
            by_effect = context.get('by_effect')
            if by_effect:
                effect_card = getattr(by_effect, 'effect_source_card', None)
                if effect_card and effect_card.owner is owner:
                    return False
            # Must have a Token or other Puppet Digimon to sacrifice
            has_sacrifice = any(
                p is not my_perm and (
                    p.is_token or
                    (p.is_digimon and p.top_card and
                     any('Puppet' in t for t in (getattr(p.top_card, 'card_traits', []) or [])))
                )
                for p in owner.battle_area
            )
            return has_sacrifice

        effect2.set_can_use_condition(condition2)

        def process2(ctx: Dict[str, Any]):
            """Delete a Token or Puppet Digimon to prevent this Digimon from leaving."""
            player = ctx.get('player')
            game = ctx.get('game')
            perm = ctx.get('permanent')
            if not (player and game and perm):
                return

            def sacrifice_filter(p):
                if p is perm:
                    return False
                if p.is_token:
                    return True
                if p.is_digimon and p.top_card:
                    traits = getattr(p.top_card, 'card_traits', []) or []
                    return any('Puppet' in t for t in traits)
                return False

            def on_sacrifice(sacrifice_perm):
                player.delete_permanent(sacrifice_perm)
                # Prevent the original permanent from leaving
                perm.will_be_remove_field = False
                if hasattr(perm, 'willBeRemoveField'):
                    perm.willBeRemoveField = False

            game.effect_select_own_permanent(
                player, on_sacrifice,
                filter_fn=sacrifice_filter,
                is_optional=True,
                prompt="Select 1 of your Tokens or [Puppet] Digimon to delete.",
            )

        effect2.set_on_process_callback(process2)
        effects.append(effect2)

        return effects

File: code/digimon-engine/tests/cards_behavioral/st19/st19_11.rs
//! ST19-11 Chaperomon.
//! Printed text covered here:
//! - [On Play] [When Digivolving] 1 opponent Digimon gets -3000 DP for the turn.
//!   If there are 3 or more Digimon, increase the DP reduction by -3000.
//! - Inherited: [All Turns] [Once Per Turn] when this Digimon would leave the
//!   battle area other than by your effects, by deleting 1 of your Tokens or
//!   other Puppet trait Digimon, prevent it from leaving.

use digimon_engine::debug_runner::{make_test_card, DebugRunner};
use digimon_engine::enums::EffectTiming;
use digimon_engine::replacement::ReplacementCause;
use digimon_engine::selection::{SelectionKind, TriggerSource};

#[test]
fn st19_11_on_play_reduces_one_opponent_digimon_by_3000_with_two_total_digimon() {
    let mut runner = DebugRunner::builder()
        .dsl_card("ST19-11")
        .expect("ST19-11 YAML loads")
        .add_card(make_digimon("OPP", 7000, &[]))
        .hand(0, &["ST19-11"])
        .memory(10)
        .start();
    runner.place_on_field(1, "OPP", Some(0));

    runner.play(0, 0).expect("play Chaperomon");
    let view = runner.pending_selection_view().expect("DP target prompt");
    assert_eq!(view.kind, SelectionKind::OppField);
    assert_eq!(
        view.valid_action_ids.len(),
        1,
        "only opponent Digimon is legal"
    );
    runner
        .execute_action(0, view.valid_action_ids[0])
        .expect("choose DP target");
    runner.auto_resolve().expect("finish DP reduction");

    let target = runner.game.players[1]
        .battle_area
        .iter()
        .position(|perm| perm.top_card().card_id(&runner.game.card_data) == "OPP")
        .expect("opponent target still present");
    let target = digimon_engine::permanent::PermanentHandle {
        player: 1,
        index: target as u8,
    };
    assert_eq!(
        runner.dp_of(target),
        Some(4000),
        "with only 2 total Digimon, target gets exactly -3000 DP"
    );
}

#[test]
fn st19_11_when_digivolving_reduces_one_opponent_digimon_by_3000_with_two_total_digimon() {
    let mut runner = DebugRunner::builder()
        .dsl_card("ST19-11")
        .expect("ST19-11 YAML loads")
        .add_card(make_digimon("BASE", 3000, &[]))
        .add_card(make_digimon("OPP", 7000, &[]))
        .start();
    let chaperomon = runner.place_stack(0, &["BASE", "ST19-11"]);
    let opp = runner.place_on_field(1, "OPP", Some(0));

    runner.game.enqueue_triggered(
        EffectTiming::WhenDigivolving,
        TriggerSource::Permanent(chaperomon),
    );
    runner.game.drain_effect_queue();

    let view = runner.pending_selection_view().expect("DP target prompt");
    assert_eq!(view.kind, SelectionKind::OppField);
    runner
        .execute_action(0, view.valid_action_ids[0])
        .expect("choose DP target");
    runner.auto_resolve().expect("finish DP reduction");

    assert_eq!(
        runner.dp_of(opp),
        Some(4000),
        "with only 2 total Digimon, target gets exactly -3000 DP"
    );
}

#[test]
#[ignore = "BLOCKED: subjectless DSL count_gte/threshold formula cannot faithfully express the 3+ Digimon extra -3000 branch"]
fn st19_11_dp_reduction_increases_to_6000_with_three_or_more_total_digimon() {
    let mut runner = DebugRunner::builder()
        .dsl_card("ST19-11")
        .expect("ST19-11 YAML loads")
        .add_card(make_digimon("BASE", 3000, &[]))
        .add_card(make_digimon("ALLY", 3000, &[]))
        .add_card(make_digimon("OPP", 7000, &[]))
        .start();
    let chaperomon = runner.place_stack(0, &["BASE", "ST19-11"]);
    runner.place_on_field(0, "ALLY", Some(0));
    let opp = runner.place_on_field(1, "OPP", Some(0));

    runner.game.enqueue_triggered(
        EffectTiming::WhenDigivolving,
        TriggerSource::Permanent(chaperomon),
    );
    runner.game.drain_effect_queue();
    let view = runner.pending_selection_view().expect("DP target prompt");
    runner
        .execute_action(0, view.valid_action_ids[0])
        .expect("choose DP target");
    runner.auto_resolve().expect("finish DP reduction");

    assert_eq!(
        runner.dp_of(opp),
        Some(1000),
        "3 total Digimon should increase the reduction to -6000"
    );
}

#[test]
fn st19_11_inherited_prevents_opponent_effect_leave_by_deleting_other_puppet() {
    let mut runner = DebugRunner::builder()
        .dsl_card("ST19-11")
        .expect("ST19-11 YAML loads")
        .add_card(make_digimon("CARRIER", 8000, &[]))
        .add_card(make_digimon("PUPPET-COST", 3000, &["Puppet"]))
        .start();
    let carrier = runner.place_stack(0, &["ST19-11", "CARRIER"]);
    runner.place_on_field(0, "PUPPET-COST", Some(0));

    runner
        .game
        .delete_permanent_with_cause(carrier, ReplacementCause::OpponentEffect);

    let accept = runner
        .pending_selection_view()
        .expect("inherited replacement accept prompt");
    assert_eq!(accept.kind, SelectionKind::Replacement);
    assert!(
        accept.is_optional,
        "prevention replacement must be a visible optional choice"
    );
    runner
        .execute_action(0, accept.valid_action_ids[0])
        .expect("accept inherited replacement");

    let view = runner
        .pending_selection_view()
        .expect("Puppet/token prevention cost prompt");
    assert_eq!(view.kind, SelectionKind::OwnField);
    assert_eq!(
        view.valid_action_ids.len(),
        1,
        "only the other Puppet, not the carrier, can pay the cost"
    );
    runner
        .execute_action(0, view.valid_action_ids[0])
        .expect("delete other Puppet");
    runner.auto_resolve().expect("finish prevention");

    assert_eq!(runner.battle_area_size(0), 1, "carrier survives");
    assert!(
        runner.game.players[0]
            .battle_area
            .iter()
            .any(|perm| perm.top_card().card_id(&runner.game.card_data) == "CARRIER"),
        "inherited carrier remains in the battle area"
    );
}

#[test]
fn st19_11_inherited_does_not_prevent_own_effect_leave() {
    let mut runner = DebugRunner::builder()
        .dsl_card("ST19-11")
        .expect("ST19-11 YAML loads")
        .add_card(make_digimon("CARRIER", 8000, &[]))
        .add_card(make_digimon("PUPPET-COST", 3000, &["Puppet"]))
        .start();
    let carrier = runner.place_stack(0, &["ST19-11", "CARRIER"]);
    runner.place_on_field(0, "PUPPET-COST", Some(0));

    runner
        .game
        .delete_permanent_with_cause(carrier, ReplacementCause::OwnEffect);

    assert!(
        runner.pending_selection_view().is_none(),
        "own effects must not offer the inherited prevention prompt"
    );
    assert_eq!(
        runner.battle_area_size(0),
        1,
        "carrier leaves; only the untouched Puppet cost remains"
    );
    assert!(
        runner.game.players[0]
            .battle_area
            .iter()
            .all(|perm| perm.top_card().card_id(&runner.game.card_data) != "CARRIER"),
        "carrier is not prevented from leaving by your own effect"
    );
}

fn make_digimon(id: &str, dp: i32, traits: &[&str]) -> digimon_engine::card_data::CardData {
    let mut card = make_test_card(id, id);
    card.dp = Some(dp);
    card.traits = traits.iter().map(|s| s.to_string()).collect();
    card
}
